src/transactions/TransactionSQL.cpp

```diff
@@ -8,14 +8,338 @@
+    XDROutputFileStream& txResultOut)
+{
+    ZoneScoped;
+    // prepare the txset for saving
+    auto lh = LedgerHeaderUtils::loadBySequence(db, sess, ledgerSeq);
+    if (!lh)
+    {
+        throw std::runtime_error("Could not find ledger");
+    }
+    auto txSet =
+        TxSetFrame::makeFromHistoryTransactions(lh->previousLedgerHash, txs);
+    TransactionHistoryEntry hist;
+    hist.ledgerSeq = ledgerSeq;
+    txSet->toXDR(hist.txSet);
+    txOut.writeOne(hist);
+    txResultOut.writeOne(results);
+}
+
+void
+writeGeneralizedTxSetToStream(Database& db, soci::session& sess,
+                              uint32 ledgerSeq,
+                              std::vector<uint8_t> const& encodedTxSet,
+                              std::vector<TransactionFrameBasePtr> const& txs,
+                              TransactionHistoryResultEntry& results,
+                              XDROutputFileStream& txOut,
+                              XDROutputFileStream& txResultOut)
+{
+    ZoneScoped;
+    UnorderedMap<Hash, TransactionFrameBase const*> txByHash;
+    for (auto const& tx : txs)
+    {
+        txByHash[tx->getFullHash()] = tx.get();
+    }
+
+    GeneralizedTxSetUnpacker unpacker(encodedTxSet, txByHash);
+    TransactionHistoryEntry hist;
+    hist.ledgerSeq = ledgerSeq;
+    hist.ext.v(1);
+    xdr_argpack_archive(unpacker, hist.ext.generalizedTxSet());
+
+    txOut.writeOne(hist);
+    txResultOut.writeOne(results);
+}
+
+// The following method handles the protocol version upgrade. Before
+// GENERALIZED_TX_SET_PROTOCOL_VERSION we don't store the tx set at
+// all, hence the results are empty. Starting from
+// GENERALIZED_TX_SET_PROTOCOL_VERSION we do store it. However, in
+// order to figure out if ledger `N` has generalized tx set we need
+// to look at ledger `N - 1` version (upgrade ledger won't have
+// generalized tx set, but will have
+// GENERALIZED_TX_SET_PROTOCOL_VERSION version).
+// Looking at ledger `N - 1` is not safe because it might be already
+// cleaned up at this point.
+// This is why we just check whatever has been loaded and skip
+// ledgers in case if the upgrade has happened in the middle of the
+// current ledger batch.
+void
+writeTxSetToStream(
+    Database& db, soci::session& sess, uint32 ledgerSeq,
+    std::vector<std::pair<uint32_t, std::vector<uint8_t>>> const& encodedTxSets,
+    std::vector<std::pair<uint32_t, std::vector<uint8_t>>>::const_iterator&
+        encodedTxSetIt,
+    std::vector<TransactionFrameBasePtr> const& txs,
+    TransactionHistoryResultEntry& results, XDROutputFileStream& txOut,
+    XDROutputFileStream& txResultOut)
+{
+    if (encodedTxSets.empty() || ledgerSeq < encodedTxSets.front().first)
+    {
+        writeNonGeneralizedTxSetToStream(db, sess, ledgerSeq, txs, results,
+                                         txOut, txResultOut);
+    }
+    else
+    {
+        // Currently there can be gaps in this method calls when there are no
```

Comment:
Is there a particular reason to keep iterating over the transactions table and handle gaps? Now that we store tx sets without gaps, would it be simpler/less error-prone to iterate over sets first and collect required transactions instead? Of course, we still need to deal with old-style tx sets, but we can extract the old-style logic into a separate function, and use the new logic on post-upgrade ledgers (asking because I spent quite a bit of time staring at this function, wondering if there's a way to make things easier to follow)

Reply:
alright, let's keep it as-is, I don't feel super strongly (also, now that the behavior has been clarified in the comments, it's easier to follow!)